extension/chrome/elements/pgp_block_modules/pgp-block-signature-module.ts

```diff
@@ -3,111 +3,101 @@
-        // ---> and user doesn't have pubkey for that email addr
-        const { pubkeys } = await this.view.pubLookup.lookupEmail(senderEmail);
-        if (!pubkeys.length) {
-          render(`Missing pubkey ${signerLongid}`, () => undefined);
-          return undefined;
-        }
-        // ---> and pubkey found on keyserver by sender email
-        const { key: pubkey } = await BrowserMsg.send.bg.await.keyMatch({ pubkeys, longid: signerLongid });
-        if (!pubkey) {
-          render(`Fetched ${pubkeys.length} sender's pubkeys but message was signed with a different key: ${signerLongid}, will not verify.`, () => undefined);
-          return undefined;
-        }
-        // ---> and longid it matches signature
-        await ContactStore.update(undefined, senderEmail, { pubkey }); // <= TOFU auto-import
-        if (retryVerification) {
-          const newResult = await retryVerification();
-          if (newResult) {
-            return newResult;
+      $('#pgp_signature').addClass('bad');
+      $('#pgp_signature > .result').text(verifyRes.error);
+      this.view.renderModule.setFrameColor('red');
+    } else if (!verifyRes || !verifyRes.signerLongids.length) {
+      $('#pgp_signature').addClass('bad');
+      $('#pgp_signature > .cursive').remove();
+      $('#pgp_signature > .result').text('Message Not Signed');
+    } else if (verifyRes.match) {
+      $('#pgp_signature').addClass('good');
+      $('#pgp_signature > .result').text('matching signature');
+    } else {
+      // is there intersection between supplied longids and signers from the message?
+      const intersection = Value.arr.intersection(verifyRes.signerLongids, verifyRes.suppliedLongids);
+      if (retryVerification) {
+        const signerEmail = this.view.getSigner();
+        if (!signerEmail) {
+          // in some tests we load the block without sender information
+          $('#pgp_signature').addClass('bad').find('.result').text(`Cannot verify: missing pubkey, missing sender info`);
+        } else {
+          this.view.renderModule.renderText('Verifying message...');
+          try {
+            const { pubkeys: fetchedPubkeys } = await this.view.pubLookup.lookupEmail(signerEmail);
+            if (fetchedPubkeys.length) {
+              for (const fetched of fetchedPubkeys) { // todo: should we ignore the ones we already have in verificationPubs?
+                // or better yet, create a common method to use both here and in ComposeStorageModule
+                ContactStore.update(undefined, signerEmail, { pubkey: fetched, pubkeyLastCheck: Date.now() }).catch(Catch.reportErr);
+              }
+              await this.renderPgpSignatureCheckResult(await retryVerification(fetchedPubkeys), fetchedPubkeys, undefined);
+              return;
+            }
+            if (intersection.length) {
+              this.renderBadSignature();
+            } else {
+              this.renderMissingPubkey(verifyRes.signerLongids[0]);
+            }
```

Comment:
I don't think you have to look for intersection here. Consider this code from iOS (which is using VerifyRes that comes from TypeScript, originally copied from browser extension, should be very similar code):

```swift
  private func renderPgpSignatureCheckResult(
    signature: VerifyRes?, 
    sender: RecipientWithOrderedPublicKeys
  ) -> ProcessedMessage.MessageSignature {
    if signature is nil then return .unsigned // render "not signed" in red
    if signature.error is not nil then return .error(signature.error) // render the "cannot verify signature: \(error)"
    if signature.signer is nil then return .unsigned // render "not signed" in red
    let signingPubKey = sender.pubkeys.find { $0.longids.contains(signature.signer) }
    if signingPubKey is nil or signature.match is nil then return .missingPubkey(longid: signature.signer) // render "cannot verify signature: no Public Key $longid" in red
    if signature.match is false then return .bad // render "bad signature" in red
    return .good
  }
```

By following the same steps, you should be able to simplify the code and retain correctness.

Please try that, see if that breaks any tests you wrote, in which case let's investigate the delta.

Reply:
You're right, I actually only need `hasIntersection` method which is implemented here 
`sender.pubkeys.find { $0.longids.contains(signature.signer) }`